core/common/src/grpc_interceptor.rs

```diff
@@ -0,0 +1,238 @@
+/// The gRPC header represents the gRPC call's Compress-Flag and Message-Length.
+const GRPC_HEADER_LENGTH: usize = 5;
+
+/// This is the trait that a gRPC Interceptor needs to imnplement.
+pub trait GrpcInterceptor: Sync {
+    /// Is this interceptor applicable?
+    ///
+    /// # Arguments
+    /// * `service_name` - The gRPC call's service name.
+    /// * `method_name` - The gRPC call's method name.
+    fn is_applicable(&self, service_name: &str, method_name: &str) -> bool;
+
+    /// Indicates that the request must be handled.
+    fn must_handle_request(&self) -> bool;
+
+    /// Indicates that the response must be handled.
+    fn must_handle_response(&self) -> bool;
+
+    /// Handle request. Return the new new request.
+    ///
+    /// # Arguments
+    /// * `service_name` - The gRPC call's service name.
+    /// * `method_name` - The gRPC call's method name.
+    /// * `protobuf_message_bytes` - The request's protobuf messages as bytes.
+    fn handle_request(
+        &self,
+        service_name: &str,
+        method_name: &str,
+        protobuf_message: Bytes,
+    ) -> Result<Bytes, Box<dyn Error + Send + Sync>>;
+
+    /// Handle response. Return the new response.
+    ///
+    /// # Arguments
+    /// * `service_name` - The gRPC call's service name.
+    /// * `method_name` - The gRPC call's method name.
+    /// * `protobuf_message_bytes` - The response's protobuf messages as bytes.
+    fn handle_response(
+        &self,
+        service_name: &str,
+        method_name: &str,
+        protobuf_message: Bytes,
+    ) -> Result<Bytes, Box<dyn Error + Send + Sync>>;
+}
+
+/// This is the type that represents the factory method for gRPC Interceptors.
+type GrpcInterceptorFactory = fn() -> Box<dyn GrpcInterceptor + Send>;
+
+/// The tower layer that hosts a service that hosts a gRPC Interceptor.
+#[derive(Clone)]
+pub struct GrpcInterceptorLayer {
+    interceptor_factory: GrpcInterceptorFactory,
+}
+
+impl GrpcInterceptorLayer {
+    /// Create the tower layer for a gRPC Interceptor.
+    ///
+    /// # Arguments
+    /// * `interceptor_factory` - The factory method for creating the desired gRPC Interceptor.
+    pub fn new(interceptor_factory: GrpcInterceptorFactory) -> Self {
+        Self { interceptor_factory }
+    }
+}
+
+impl<S> Layer<S> for GrpcInterceptorLayer {
+    type Service = GrpcInterceptorService<S>;
+
+    fn layer(&self, service: S) -> Self::Service {
+        GrpcInterceptorService { service, interceptor_factory: self.interceptor_factory }
+    }
+}
+
+/// The tower service that hosts a gRPC Interceptor.
+#[derive(Clone)]
+pub struct GrpcInterceptorService<S> {
+    service: S,
+    interceptor_factory: GrpcInterceptorFactory,
+}
+
+impl<S> GrpcInterceptorService<S> {
+    /// Retrieve the gRPC service name and method name from a URI.
+    ///
+    /// * `uri` - The uri used for the gRPC call.
+    fn retrieve_grpc_names_from_uri(uri: &Uri) -> (String, String) {
+        let uri_string = uri.to_string();
```

Comment:
can you document the expected format of a uri here?

this also might be a good place to consider a regex with capture groups if the format isn't horribly complex

Reply:
I provided an example and switched over to using regex.